Content.Server/Implants/SubdermalImplantSystem.cs

```diff
@@ -107,42 +111,84 @@ private void OnScramImplant(EntityUid uid, SubdermalImplantComponent component,
 
         var xform = Transform(ent);
-        var entityCoords = xform.Coordinates.ToMap(EntityManager, _xform);
+        var targetCoords = SelectRandomTileInRange(xform, implant.TeleportRadius);
 
-        // try to find a valid position to teleport to, teleport to whatever works if we can't
-        var targetCoords = new MapCoordinates();
-        for (var i = 0; i < implant.TeleportAttempts; i++)
+        if (targetCoords != null)
         {
-            var distance = implant.TeleportRadius * MathF.Sqrt(_random.NextFloat()); // to get an uniform distribution
-            targetCoords = entityCoords.Offset(_random.NextAngle().ToVec() * distance);
+            _xform.SetWorldPosition(ent, targetCoords.Value.Position);
+            _xform.AttachToGridOrMap(ent, xform);
+            _audio.PlayPvs(implant.TeleportSound, ent);
+            args.Handled = true;
+        }
+    }
 
-            // prefer teleporting to grids
-            if (!_mapManager.TryFindGridAt(targetCoords, out var gridUid, out var grid))
-                continue;
+    private MapCoordinates? SelectRandomTileInRange(TransformComponent userXform, float radius)
+    {
+        var userCoords = userXform.Coordinates.ToMap(EntityManager, _xform);
+        var grids = _lookupSystem.GetEntitiesInRange<MapGridComponent>(userCoords, radius).ToList();
+        _random.Shuffle(grids);
 
-            // the implant user probably does not want to be in your walls
-            var valid = true;
-            foreach (var entity in grid.GetAnchoredEntities(targetCoords))
+        // Give preference to the grid the entity is currently on.
+        var idx = grids.FindIndex(grid => grid.Owner == userXform.GridUid);
+        if (idx != -1)
+        {
+            if (_random.Prob(0.66f))
+            {
+                (grids[0], grids[idx]) = (grids[idx], grids[0]);
+            }
+            else
             {
-                if (!_physicsQuery.TryGetComponent(entity, out var body))
-                    continue;
+                (grids[^1], grids[idx]) = (grids[idx], grids[^1]);
+            }
+        }
 
-                if (body.BodyType != BodyType.Static ||
-                    !body.Hard ||
-                    (body.CollisionLayer & (int) CollisionGroup.Impassable) == 0)
-                    continue;
+        MapCoordinates? targetCoords = null;
 
-                valid = false;
-                break;
+        foreach (var grid in grids)
+        {
+            var valid = false;
+
+            var range = (float) Math.Sqrt(radius);
+            var box = Box2.CenteredAround(userCoords.Position, new Vector2(range, range));
+            var tilesInRange = _mapSystem.GetTilesEnumerator(grid.Owner, grid.Comp, box, false);
+            var tileList = new List<TileRef>();
+
+            while (tilesInRange.MoveNext(out var tile))
+            {
+                tileList.Add(tile);
             }
+
+            _random.Shuffle(tileList);
+
```

Comment:
Take is much faster because it only moves 2 entries instead of the entire thing.